sharding/src/node/shard.rs

```diff
@@ -75,46 +75,175 @@ impl<'a> Shard<'a> {
+        &mut self,
         listener: Arc<RwLock<TcpListener>>,
         rw_stream: Arc<RwLock<Option<TcpStream>>>,
     ) {
         let listener_guard = listener.read().unwrap();
         match listener_guard.accept() {
             Ok((stream, addr)) => {
                 println!("New connection accepted from {}.", addr);
-                let mut stream_guard = rw_stream.write().unwrap();
-                *stream_guard = Some(stream);
+                // let mut stream_guard = rw_stream.write().unwrap();
+
+                self.router_stream = Arc::new(RwLock::new(Some(stream)));
+
+                // Start listening for incoming messages in a thread
+                let mut shard_clone = self.clone();
+                let _handle = thread::spawn(move || {
+                    shard_clone.listen();
+                });
             }
             Err(e) => {
                 eprintln!("Failed to accept a connection: {}", e);
             }
         }
     }
 
-    pub fn update(&mut self) -> Result<(), io::Error> {
-        self.memory_manager.update()
+    // Listen for incoming messages
+    pub fn listen(&mut self) {
+        println!("Listening for incoming messages");
+        let self_clone = self.clone();
+        let stream = self_clone.router_stream.read().unwrap();
+        loop {
+            // TODO-SHARD fix this multiple match statements so they're not nested
+            match stream.as_ref() {
+                Some(mut stream) => {
+                    let mut buffer = [0; 1024];
+                    let mut retries = 10;
+                    match stream.read(&mut buffer) {
+                        Ok(_) => {
+                            let message_string = String::from_utf8_lossy(&buffer);
+                            // println!("Received message: {}", message_string);
+                            match self.get_response_message(&message_string) {
+                                Some(response) => {
+                                    stream.write(response.as_bytes()).unwrap();
+                                }
+                                None => {
+                                    retries -= 1;
+                                    if retries == 0 {
+                                        eprintln!("Stream has exceeded the number of retries");
+                                        break;
+                                    }
+                                }
+                            }
+                        }
+                        Err(e) => {
+                            eprintln!("Failed to read from stream: {:?}", e);
+                        }
+                    }
+                }
+                None => {
+                    // Do nothing
+                }
+            };
+        }
+    }
+
+    fn get_response_message(&mut self, message: &str) -> Option<String> {
+        let message = match Message::from_string(&message) {
+            Ok(message) => message,
+            Err(e) => {
```

Comment:
`e` is unused if the print is commented
```suggestion
            Err(_e) => {
```